### ISSUE TITLE
unable to build from master, cyclic dependency between lib-math and lib-project-rate

### ISSUE THREAD

**Issue by avoidik** [audacity/audacity#1668]
hello,

while trying to build from the master I've got a cyclic dependency error

```bash
$ docker build -t audacity_linux_env linux/ubuntu-focal
$ docker run --rm -v $(pwd):/audacity/audacity/ -v $(pwd)/build/linux-system:/audacity/build -it audacity_linux_env
```

error details

```
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "lib-math" of type SHARED_LIBRARY
    depends on "lib-project-rate" (weak)
  "lib-project-rate" of type SHARED_LIBRARY
    depends on "lib-math" (weak)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

cheers

**Comment by Paul-Licameli** [comment]
Project rate should depend on math but not the opposite.

**Comment by Paul-Licameli** [comment]
https://github.com/audacity/audacity/blob/master/libraries/lib-math/CMakeLists.txt

There is no dependency on the lib-project-rate target there.  I don’t get this build error.

**Comment by Paul-Licameli** [comment]
Did you make modifications?

**Comment by IshoAntar** [comment]
@Paul-Licameli I can confirm this, although not in Audacity (!). It happens on my fork too when i cherry-picked that commit.

> Project rate should depend on math but not the opposite.

The dependency graph confirms this:

![modules](https://user-images.githubusercontent.com/32076082/133346958-edec6351-2b05-40a6-b32f-8446a8eea233.png)

Thought it might be a caching issue but doing a clean build didn't help. Still investigating the issue.

**Comment by IshoAntar** [comment]
@Paul-Licameli I can confirm this is an issue in #1588. It happened in my fork too when i cherry-picked that (merge) commit ([workflow run](https://github.com/IshoAntar/AudMonkey/runs/3605993959?check_suite_focus=true)), happens locally too.

No idea what could be the issue as the other builds (in the same workflow) is succeeding?!

**Comment by avoidik** [comment]
I was using build instructions for Docker from original repository, with minor changes to path and inline variables, you may try to reproduce it by following these two steps

```bash
$ docker build -t audacity_linux_env linux/ubuntu-focal
```

```bash
$ docker run --rm -v $(pwd):/audacity/audacity/ -v $(pwd)/build/linux-system:/audacity/build -it audacity_linux_env
```

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli I can confirm this is an issue in #1588. It happened in my fork too when i cherry-picked that (merge) commit ([workflow run](https://github.com/IshoAntar/AudMonkey/runs/3605993959?check_suite_focus=true)), happens locally too.
> 
> No idea what could be the issue as the other builds (in the same workflow) are succeeding?!

Hello, esteemed Mr. Monkey.  Are you sure the strongly connected component in the error message is still the same cycle of only two targets?

This is weird.  I am known to be very alert to the evils of cyclic dependencies and I avoid making them.  Levelizing the code better is the point of this initiative to extract libraries.

I think I recall something in CMake Docs saying that cycles can be tolerated when the libraries are statically linked.

**Comment by IshoAntar** [comment]
> Hello, esteemed Mr. Monkey.

Hey there!

> Are you sure the strongly connected component in the error message is still the same cycle of only two targets?

No. it's the cycle of 0 targets:

![modules](https://user-images.githubusercontent.com/32076082/133511008-fbf733db-c1fc-4935-b927-055acd5184a1.png)

`lib-project-rate` doesn't depend on anything except the main target.

> This is weird.

It is indeed. This is the first time that i encountered this sort of thing.

> I think I recall something in CMake Docs saying that cycles can be tolerated when the libraries are statically linked.

Static linking would be an option. Any downsides?

**Comment by Paul-Licameli** [comment]
> > Hello, esteemed Mr. Monkey.
> 
> Hey there!
> 
> > Are you sure the strongly connected component in the error message is still the same cycle of only two targets?
> 
> No. it's the cycle of zero targets:
> 
> ![modules](https://user-images.githubusercontent.com/32076082/133511008-fbf733db-c1fc-4935-b927-055acd5184a1.png)
> 
> > This is weird.
> 
> It is indeed. This is the first time that i encountered this sort of thing.
> 
> > I think I recall something in CMake Docs saying that cycles can be tolerated when the libraries are statically linked.
> 
> Static linking would be an option. Any downsides?

Good for you, that you pay attention to the dependency graph generated in the build!  But a few things.

* You are not seeing the result of these CMake options: https://cmake.org/cmake/help/v3.9/module/CMakeGraphVizOptions.html
* You are seeing the results of some code I wrote myself in CMake language, which gives me more control over Graphviz styles, colors, shapes, etc.
* So it’s not really a picture of what CMake thinks the dependencies are.
* Could there be bugs in CMake itself?  Do we need to identify a minimum version of CMake without those bugs?
* Finally, you are reading the depends-on relation backwards.  The extension modules (rectangles) depend on the executable (pentagon) which depends on third party libraries (ellipses) and the project’s own libraries (octagons).

**Comment by IshoAntar** [comment]
> You are not seeing the result of these CMake options: cmake.org/cmake/help/v3.9/module/CMakeGraphVizOptions.html

I was today years old when i learned that CMake has built-in support for GraphViz!

> You are seeing the results of some code I wrote myself in CMake language, which gives me more control over Graphviz styles, colors, shapes, etc.
So it’s not really a picture of what CMake thinks the dependencies are.

Still, reading through the scripts, i don't see how `lib-math` depends on `lib-project-rate`? There is nothing explicitly adding the dependency.

> Could there be bugs in CMake itself?

That's a possibility

> Do we need to identify a minimum version of CMake without those bugs?

I'm pretty sure it's not related to a specific version. GitHub's Ubuntu runner [is using CMake 3.21.2](https://github.com/actions/virtual-environments/blob/ubuntu18/20210906.1/images/linux/Ubuntu1804-README.md#tools), the same version in the Arch container and my local installation, Ubuntu is building successfully while Arch (local and CI) is failing.

Might be that different distros compile there own versions of the packages and some compiler switch/configuration is triggering a bug in CMake?

> Finally, you are reading the depends-on relation backwards. The extension modules (rectangles) depend on the executable (pentagon) which depends on third party libraries (ellipses) and the project’s own libraries (octagons).

Oops! Thanks for the heads up.

If the problem persists, wouldn't it be possible to build one monolithic library (e.g. `libaudacity.so`) that statically links against all of these tiny libraries and use that in the rest of the code base?

**Comment by avoidik** [comment]
using latest cmake 3.21.2 doesn't change much, but if I change this I can get through the error

```diff
diff --git a/cmake-proxies/cmake-modules/AudacityFunctions.cmake b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
index 1f4d436..63ebe5e 100644
--- a/cmake-proxies/cmake-modules/AudacityFunctions.cmake
+++ b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
@@ -532,7 +532,7 @@ macro( audacity_library NAME SOURCES IMPORT_TARGETS
       "${IMPORT_TARGETS}"
       "${ADDITIONAL_DEFINES}"
       "${ADDITIONAL_LIBRARIES}"
-      "SHARED"
+      "STATIC"
    )
    set( GRAPH_EDGES "${GRAPH_EDGES}" PARENT_SCOPE )
    # Collect list of libraries for the executable to declare dependency on
```

**Comment by Paul-Licameli** [comment]
> using the latest cmake 3.21.2 doesn't change much, but if I change this I can get through the error
> 
> ```diff
> diff --git a/cmake-proxies/cmake-modules/AudacityFunctions.cmake b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
> index 1f4d436..63ebe5e 100644
> --- a/cmake-proxies/cmake-modules/AudacityFunctions.cmake
> +++ b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
> @@ -532,7 +532,7 @@ macro( audacity_library NAME SOURCES IMPORT_TARGETS
>        "${IMPORT_TARGETS}"
>        "${ADDITIONAL_DEFINES}"
>        "${ADDITIONAL_LIBRARIES}"
> -      "SHARED"
> +      "STATIC"
>     )
>     set( GRAPH_EDGES "${GRAPH_EDGES}" PARENT_SCOPE )
>     # Collect list of libraries for the executable to declare dependency on
> ```
> 
> I assume that `audacity_library` macro might be extended to include 6th parameter to specifically define are we're handling shared or statically linked library

I’m not sure that the change to static linking would be free of bad consequences.  Are there global variables that would be replicated, for example?

**Comment by crsib** [comment]
This issue was fixed a while ago in #1787
